(WIP) FormBuilder: settings for token timeout and redirection url

aydun opened the pull request:
Overview
----------------------------------------
Allow authentication token validity period to be set on form and redirect to an alternative URL if authentication token fails.

Create a FormBuilder form.  
Tick the 'Provide Email Token'.  
Create an email and include the token.

Before
----------------------------------------
All tokens use the system default setting for checksums.
If a user clicks on a URL with an expired token (for example in an old email), they get a terse error message.

After
----------------------------------------
If the 'Provide Email Token' is ticked in FormBuilder UI, 2 new fields appear:
- Token Timeout: specifies the number of days the token will be valid for.
- Token Redirect: specifies a URL to redirect to if the authentication fails - for example, if the token is expired.  This URL could be a simple page with instructions about how to get a new URL or another form suitable for unauthenticated user.

The token will include `_authxRedir=civicrm/blah` parameter.  If authentication fails, the user is redirected to that page.  The error message is passed as a parameter to that page and could be used to explain why the user has been redirected.

The token will use the specified timeout if set, otherwise the system-wide value.

Technical Details
----------------------------------------
New `authx_redirect` and `authx_timeout` fields on afform entity and FormBuilder UI


Comments
----------------------------------------

aydun committed:
Enable redirection if authx authentication fails (eg expired token)

aydun committed:
Add 'authx_redirect' field to afform

aydun committed:
Only add '_authxRedir' if param is set

aydun committed:
Add Token Timeout - allows validity period to be set per form
Move 'Provide Email Token' and related to end of the 'Placement' section
- seems to flow better.
Make 'Provide Email Token' show as disabled instead of hidden if no page
route
Change Token construciton to use new form value

colemanw commented on ext/afform/core/CRM/Afform/AfformScanner.php:
```diff
@@ -149,6 +149,8 @@ public function getMeta($name) {
       'is_dashlet' => FALSE,
       'is_public' => FALSE,
       'is_token' => FALSE,
+      'authx_timeout' => 0,
```
This defaults to `0` but in `Civi\Afform\Tokens` it uses the `??` operator to check if this is set. So I think it should default to `NULL`.

aydun replied:
Thanks, fixed

aydun committed:
Fix default

aydun committed:
Restrict Token Timeout to SuperUsers

colemanw commented on ext/afform/core/Civi/Api4/Afform.php:
```diff
@@ -168,6 +168,13 @@ public static function getFields($checkPermissions = TRUE) {
+        [
+          'name' => 'authx_timeout',
+          'data_type' => 'Integer',
+        ],
```
We also need to enforce permissions at the api level. I think that because Afform is a nonstandard API this metadata won't automatically work, but it's still semantically correct to add it:
```suggestion
          'permission' => ['all CiviCRM permissions and ACLs'],
        ],
```
Then I guess we need to add some handling for this in the Afform::save api, e.g. loop through `getFields` and unset any values the user doesn't have permission to edit prior to saving (and prior to merging values with the existing form during update, so that the user isn't able to change an exiting value)

aydun committed:
Update ext/afform/core/Civi/Api4/Afform.php

Co-authored-by: colemanw <coleman@civicrm.org>

totten committed:
(NFC) _authx_reload() - Add comment

If we're now accepting editable parameters for the redirect target, then we
need to be sure it's not an external domain.

totten commented on ext/authx/Civi/Authx/Authenticator.php:
```diff
@@ -190,6 +194,9 @@ protected function checkCredential($tgt) {
 
     if ($checkEvent->getRejection()) {
+      if ($tgt->redirect) {
+        _authx_reload($tgt->redirect, 'msg="' . $checkEvent->getRejection() . '"');
```
I was concerned about this becoming vector for an "open redirect" (https://learn.snyk.io/lesson/open-redirect/). But I think (through a bit of happenstance) it turns out OK as-is. Specifically, if you trace `_authx_reload()`, the `$redirect` is passed through `CRM_Utils_System::url()`:

* If `url()` runs in the default mode (`$absolute==FALSE`), then you could give a malicious value (`/example.com/foobar`) which expands to a legal/external URL (`//example.com/foobar`  -- ie `example.com` with an inherited scheme, `http:` or `https:`).
    * (_This may depend on the usage of clean URLs. My local dmaster behaved this way, but my local wpmaster didn't._)
* In actuality, `url()` is called with `$absolute==TRUE`. In this mode, I don't see a way to trick it into an external redirect.
    * (_I don't remember any particular reason for this flag -- just a feeling that absolute URLs are vaguely more reliable._)
    * (_This is basically what `snyk.io` refers to as "Implement a fixed domain"._)

I can imagine someone innocently flipping that to insecure mode... so I pushed up a small comment to explain that it's now important.

eileenmcnaughton commented:
@aydun @totten @ufundo this is stale now - but I'm wondering how it fits with interim changes with authentication?

aydun commented:
@eileenmcnaughton I've changed it to WIP and will review

colemanw commented:
This seems somewhat related to @ufundo's work on https://github.com/civicrm/civicrm-core/pull/32965 and perhaps the two would complement each other?

colemanw commented:
@aydun I have rebased this as https://github.com/civicrm/civicrm-core/pull/36291

aydun commented:
Thanks @colemanw
